### PR TITLE
[pytorch][PR] Functional conv2d

### PR DESCRIPTION
**Pull request by z-a-f**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#21225 [pytorch][PR] Functional conv2d**&nbsp;&nbsp;[:green_heart:](https://our.internmc.facebook.com/intern/diff/D15544061/)

Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:white_circle:&nbsp; https://github.com/pytorch/pytorch/issues/21323 Quantized Conv2d Module&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D15551835/)
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **https://github.com/pytorch/pytorch/issues/21225 Functional conv2d**&nbsp;&nbsp;[:yellow_heart:](https://our.intern.facebook.com/intern/diff/D15544061/)

Pull Request resolved: https://github.com/pytorch/pytorch/pull/21225

Differential Revision: [D15544061](https://our.internmc.facebook.com/intern/diff/D15544061/)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 1]
```diff
@@ -22,21 +22,19 @@ class QConv2dInt8 final : public c10::OperatorKernel {
...
-    TORCH_CHECK(dilation.size() == 2, "2D convolution only");
-    TORCH_CHECK(output_padding.size() == 2, "2D convolution only");
-    TORCH_CHECK(
-        (dilation[0] == 1 && dilation[1] == 1),
-        "Currently dilation should be 1");
-    TORCH_CHECK(
-        (output_padding[0] == 0 && output_padding[1] == 0),
-        "Currently output padding should be 0");
+    TORCH_CHECK(fbgemm::fbgemmSupportedCPU(),
+                "[QConv2dInt8] Your CPU does not support FBGEMM.");
```
Comment (dzhulgakov) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp]: is this prefix for unittesting only? feels a bit sketchy to users. Can't we use stacktrace for testing?

Comment (zafartahirov) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Yes, it is only for unittesting -- to make sure the errors are raised. The problem is that cpp raises `RuntimeError` no matter what error it is. I'm not sure how to make sure the correct error was raised except putting a flag in the TORCH_CHECK

Comment (zafartahirov) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp]: Removed it -- using a `np.assert_raise_regex`

File: torch/nn/quantized/functional.py [hunk 2]
```diff
@@ -1,4 +1,82 @@
...
+           scale=1.0, zero_point=0,
+           dtype=torch.quint8,
+           prepacked=True):
+    r"""
+    conv2d(input, weight, bias,
+           stride=1, padding=0, dilation=1, groups=1,
+           padding_mode='zeros',
+           scale=1.0, zero_point=0,
+           dtype=torch.quint8,
+           prepacked=True) -> Tensor
```
Comment (raghuramank100) [inline torch/nn/quantized/functional.py]: Suggest having functionals only take in unpacked weights and do packing internally. Remove prepacked option for functional.

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Please see included comments

**Review by raghuramank100 — APPROVED** [review]
Approved, need to remove support for packed weights in a later diff.

**Pull request closed without merging**